test/node.go

```diff
@@ -0,0 +1,567 @@
+	}
+	return n.Tracker.AwaitTransactions(ctx, sentHashes)
+}
+
+// AwaitSentTransactions awaits all the transactions that this node has sent
+// via SendCelo.
+func (n *Node) AwaitSentTransactions(ctx context.Context) error {
+	return n.AwaitTransactions(ctx, n.SentTxs...)
+}
+
+// ProcessedTxBlock returns the block that the given transaction was processed
+// in, nil will be retuned if the transaction has not been processed by this node.
+func (n *Node) ProcessedTxBlock(tx *types.Transaction) *types.Block {
+	return n.Tracker.GetProcessedBlock(tx.Hash())
+}
+
+// TxFee returns the gas fee for the given transaction.
+func (n *Node) TxFee(ctx context.Context, tx *types.Transaction) (*big.Int, error) {
+	r, err := n.WsClient.TransactionReceipt(ctx, tx.Hash())
+	if err != nil {
+		return nil, err
+	}
+	return big.NewInt(0).Mul(new(big.Int).SetUint64(r.GasUsed), tx.GasPrice()), nil
+}
+
+// Network represents a network of nodes and provides functionality to easily
+// create, start and stop a collection of nodes.
+type Network []*Node
+
+type NodeConfig struct {
+	ValidatorAccount *env.Account
+	DevAccount       *env.Account
+	*node.Config
+}
+
+func NewNodeConfig(validatorAccount, devAccount *env.Account) *NodeConfig {
+	confCopy := *baseNodeConfig
+	return &NodeConfig{
+		ValidatorAccount: validatorAccount,
+		DevAccount:       devAccount,
+		Config:           &confCopy,
+	}
+}
+
+func Accounts(numValidators int) *env.AccountsConfig {
+	return &env.AccountsConfig{
+		Mnemonic:             env.MustNewMnemonic(),
+		NumValidators:        numValidators,
+		ValidatorsPerGroup:   1,
+		NumDeveloperAccounts: numValidators,
+	}
+}
+
+func GenesisConfig(accounts *env.AccountsConfig) *genesis.Config {
+	gc := genesis.CreateCommonGenesisConfig(
+		big.NewInt(1),
+		accounts.AdminAccount().Address,
+		params.IstanbulConfig{
+			Epoch:          10,
+			ProposerPolicy: uint64(istanbul.ShuffledRoundRobin),
+			LookbackWindow: 3,
+			BlockPeriod:    0,
+
+			// 50ms is a really low timeout, we set this here because nodes
+			// fail the first round of consensus and so setting this higher
+			// makes tests run slower.
+			RequestTimeout: 50,
+		},
+	)
+	genesis.FundAccounts(gc, accounts.DeveloperAccounts())
+	return gc
+}
+
+// NewNetwork generates a network of nodes that are running and mining. For
+// each provided validator account a corresponding node is created and each
+// node is also assigned a developer account, there must be at least as many
+// developer accounts provided as validator accounts. If there is an error it
+// will be returned immediately, meaning that some nodes may be running and
+// others not.
+func NewNetwork(accounts *env.AccountsConfig, gc *genesis.Config) (Network, error) {
+
+	genesis, err := genesis.GenerateGenesis(accounts, gc, "../monorepo/packages/protocol/build/contracts")
+	if err != nil {
+		return nil, err
+	}
+
+	validatorAccounts := accounts.ValidatorAccounts()
+	network := make([]*Node, len(validatorAccounts))
```

Comment:
More a note than nit, but I usually set the capacity with make (`make(T, 0, (len(validatorAccounts))`) and then use append rather than directly setting the value.

Reply:
I do that when I don't have access to an index variable, for example copying from a map to a pre-allocated slice. But when I have access to the index I think prefer this approach, just because its slightly more concise and I find it more explicit about the fact that only the pre-allocated memory is being used.

```
network[i] = n
// vs
network = append(network, n)
```